### PR TITLE
Refactored header name matching for WASM target

### PR DESCRIPTION
**Pull request by matiu2**
Advantages:

 1. Shorter code
 2. More rusty and less macroy
 3. It should work in wasm now untested and make yew users happier
    https://github.com/DenisKolodin/yew/issues/478

(Un?)fortunately I also ran rustfmt on it - which changed the layouts for the tables a bit but not hideously.

### CODE REVIEW

**Comment by carllerche** [comment]
Hi, thanks for the work.

Would it be possible to include benchmarks for parsing header names? Probably could add a bench parsing each one + some common unknown header names. It would be interesting to compare the before and after.

**Comment by matiu2** [comment]
I tried to make it work in nightly and beta, but then it broke for rust 1.20.

The existing benches are pretty much the same, for example:

New: `test basic::insert_500_custom_headers::header_map            ... bench:      35,208 ns/iter (+/- 3,372)`

Old: `test basic::insert_500_custom_headers::header_map            ... bench:      36,283 ns/iter (+/- 573)`

I'll some benches just for the name parsing..

**Comment by matiu2** [comment]
I don't know how to bench unless it's running rust nightly (cargo bench) - so I had to make it work in nightly to get the numbers. It seems pretty close:

`cargo bench header_name`

# old macro style
test header_name_bad     ... bench:          15 ns/iter (+/- 0)
test header_name_easy    ... bench:          16 ns/iter (+/- 2)
test header_name_various ... bench:      10,222 ns/iter (+/- 991)

# new giant match statement
test header_name_bad     ... bench:          20 ns/iter (+/- 1)
test header_name_easy    ... bench:          21 ns/iter (+/- 1)
test header_name_various ... bench:       9,909 ns/iter (+/- 1,133)

File: benches/header_name.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    b.iter(|| {
+        HeaderName::from_bytes(&name[..]).expect_err("Bad header name");
+    });
+}
+
+#[bench]
+fn header_name_various(b: &mut Bencher) {
+
+    b.iter(|| {
+        let all_known_headers = make_all_known_headers();
```
Comment (seanmonstar) [inline benches/header_name.rs]: This could be lifted out of the `iter`, so it isn't part of the measurements.

**Comment by matiu2** [comment]
I've moved the make_all_known_headers out of the iterator. Sorry for not noticing that.

This commit is definitely slowing the code down by about 6% on my laptop :(

# New bench mark results

## Old code:

test header_name_bad     ... bench:          15 ns/iter (+/- 11)
test header_name_easy    ... bench:          16 ns/iter (+/- 1)
test header_name_various ... bench:       3,689 ns/iter (+/- 290)

## New code:

test header_name_bad     ... bench:          20 ns/iter (+/- 0)
test header_name_easy    ... bench:          24 ns/iter (+/- 1)
test header_name_various ... bench:       3,921 ns/iter (+/- 395)

--------

The main reason for this commit is so we can use `http` in wasm without an optimized build.

When debug building, those macros that expand out to `...&& b[0 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1] == b'p'...` get put as-is into wasm, causing the wasm implementiation to blow up at run-time trying to add all the ones.

Here's the issue in `yew` that I'm trying to fix: https://github.com/DenisKolodin/yew/issues/478#issuecomment-507429151

**Comment by matiu2** [comment]
Still slow. I don't know why :(

# Latest bench mark results

## Old macro based code:

test header_name_bad     ... bench:          15 ns/iter (+/- 0)
test header_name_easy    ... bench:          16 ns/iter (+/- 0)
test header_name_various ... bench:       3,648 ns/iter (+/- 123)

## New giant match statement code:

test header_name_bad     ... bench:          25 ns/iter (+/- 0)
test header_name_easy    ... bench:          24 ns/iter (+/- 0)
test header_name_various ... bench:       4,568 ns/iter (+/- 153)

**Comment by matiu2** [comment]
Telling it the len of the array makes it a bit faster:

test header_name_bad     ... bench:          25 ns/iter (+/- 0)
test header_name_easy    ... bench:          24 ns/iter (+/- 0)
test header_name_various ... bench:       4,566 ns/iter (+/- 206)

**Comment by matiu2** [comment]
# Latest bench mark results

## Old macro based code:

    $ cargo bench header_name
    ...
    test header_name_bad     ... bench:          14 ns/iter (+/- 0)
    test header_name_easy    ... bench:          16 ns/iter (+/- 1)
    test header_name_various ... bench:       3,763 ns/iter (+/- 80)

## New giant match statement code:

    test header_name_bad     ... bench:          25 ns/iter (+/- 1)
    test header_name_easy    ... bench:          24 ns/iter (+/- 0)
    test header_name_various ... bench:       4,227 ns/iter (+/- 75)

**Comment by carllerche** [comment]
We should definitely get this working with wasm. Perhaps there is a way to conditional compile a different fn in wasm?

**Comment by matiu2** [comment]
This version uses conditional compilation. It'll only compile the wasm version if there are 'debug_assertions' and the target is 'wasm32'.

It make the code longer, and harder to maintain though, because now if you add a new header, you'll have to add it in two places.

I think simplifying the code to the giant match statement is best. Although 8 more nanoseconds per header check may seem a lot; in msecs it is 8e-6 (0.0000008 msecs). But I'll leave the decision in your capable hands.

I tested it and it works in my wasm app in debug mode :)

**Pull request merged**